### PR TITLE
[SPARK-50463][SQL] Fix `ConstantColumnVector` with Columnar to Row conversion

### PR DESCRIPTION
**Pull request by richardc-db**
### What changes were proposed in this pull request?

https://github.com/apache/spark/commit/800faf0abfa368ad0a5ef1e0fa44b74dbaab724e frees column vector resources between batches in columnar to row conversion. However, like `WritableColumnVector`, `ConstantColumnVector` should not free resources between batches because the same data is used across batches

### Why are the changes needed?

Without this change, ConstantColumnVectors with string values, for example, will fail if used with column->row conversion. For instance, reading a parquet table partitioned by a string column with multiple batches.

### How was this patch tested?

added UT that failed before and now passes

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by richardc-db** [comment]
hey @viirya mind taking a quick look at this one? Fixes using `ConstantColumnVector`s with the col->row memory leak fix you merged a bit ago

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ConstantColumnVector.java [hunk 1]
```diff
@@ -77,6 +77,11 @@ public ConstantColumnVector(int numRows, DataType type) {
     }
   }
 
+  public void closeIfFreeable() {
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ConstantColumnVector.java:80]: `closeIfNotWritable`?

**Review by viirya — DISMISSED** [review]
Good catch. Thanks.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ConstantColumnVector.java (hunk 1 above)
Comment (richardc-db) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ConstantColumnVector.java:80]: ah yeah my apologies - do you think `closeIfNotWritable` still makes sense given that the `ConstantColumnVector` is not a `WritableColumnVector`? 

Do you think `closeIfFreeable` makes sense? If so, I'll change the method name in `ColumnVector.java` in this PR as well. 

Or do you have a preference for a different name?

File: sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java [hunk 2]
```diff
@@ -69,14 +69,14 @@ public abstract class ColumnVector implements AutoCloseable {
   public abstract void close();
 
   /**
-   * Cleans up memory for this column vector if it's not writable. The column vector is not usable
-   * after this.
+   * Cleans up memory for this column vector if it's resources are freeable bewteen batches.
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:72]:
```suggestion
   * Cleans up memory for this column vector if it's resources are freeable between batches.
```

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java [hunk 3]
```diff
@@ -69,14 +69,14 @@ public abstract class ColumnVector implements AutoCloseable {
...
+   * Cleans up memory for this column vector if it's resources are freeable between batches.
+   * The column vector is not usable after this.
    *
-   * If this is a writable column vector, it is a no-op.
+   * If this is a writable column vector or constant column vector, it is a no-op.
    */
-  public void closeIfNotWritable() {
-    // By default, we just call close() for all column vectors. If a column vector is writable, it
-    // should override this method and do nothing.
+  public void closeIfFreeable() {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:77]: Hmm `ColumnVector` is an API actually. Can we avoid direct renaming here

Comment (richardc-db) [reply sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:77]: The original PR which added this method https://github.com/apache/spark/commit/800faf0abfa368ad0a5ef1e0fa44b74dbaab724e was merged ~a month ago. is it too late to now change the method name?

If so, how do you suggest we go about this? I think `closeIfNotWritable` doesn't represent the purpose of the method well, so one option is to deprecate `closeIfNotWritable` and directly call `closeIfFreeable` inside it. This is pretty ugly, though, so I'm open to any suggestions

Comment (HyukjinKwon) [reply sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:77]: ah okie dokie then should be fine

Comment (viirya) [reply sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java:77]: Yea, it is a new API just added recently and not released yet.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @richardc-db , @viirya , @HyukjinKwon .
Merged to master for Apache Spark 4.0.0 on February 2025.

**Comment by LuciferYang** [comment]
@dongjoon-hyun I found that the previous PR of this one, which is https://github.com/apache/spark/pull/48767, was merged into both master and branch-3.5, but this pr was only merged into master. Then I manually copied the new test cases from this pr and found that they fail in branch-3.5. 

So should we backport this patch to branch-3.5? I have created a backporting: https://github.com/apache/spark/pull/49131

**Comment by dongjoon-hyun** [comment]
Oh, is this a regression then, @LuciferYang ?

**Comment by LuciferYang** [comment]
> Oh, is this a regression then, @LuciferYang ?

I think so...

**Pull request closed without merging**